lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -0,0 +1,270 @@
+
+#include <google/protobuf/repeated_field.h>
+#include <google/protobuf/timestamp.pb.h>
+#include <google/protobuf/util/time_util.h>
+#include <grpcpp/channel.h>
+#include "magma_logging.h"
+#include "EnumToString.h"
+#include "SessionStateEnforcer.h"
+
+namespace magma
+{
+//temp routine
+void call_back_void_upf(grpc::Status, magma::UpfRes response)
+{
+  //do nothinf but to only passing call back
+  //cout <<" Only for testing call back" << endl;
+}
+
+
+/*constructor*/
+SessionStateEnforcer::SessionStateEnforcer(
+    std::shared_ptr<StaticRuleStore> rule_store,
+    SessionStore& session_store,
+    std::shared_ptr<PipelinedClient> pipelined_client,
+    std::shared_ptr<AmfServiceClient> amf_srv_client,
+    magma::mconfig::SessionD mconfig)
+    : session_store_(session_store),
+      pipelined_client_(pipelined_client),
+      amf_srv_client_(amf_srv_client),
+      retry_timeout_(1),
+      mconfig_(mconfig)
+      //SmSession_Context_Send_Client_sp_(SmSession_Context_Send_Client)
+      //pipelined_client_(pipelined_client),
+      {
+	  // for now this is the right place, need to move if find  anohter right place
+          static_rule_init();
+      }
+
+void SessionStateEnforcer::attachEventBase(folly::EventBase* evb) {
+   evb_ = evb;
+}
+
+void SessionStateEnforcer::stop() {
+   evb_->terminateLoopSoon();
+}
+
+folly::EventBase& SessionStateEnforcer::get_event_base() {
+   return *evb_;
+}
+
+bool SessionStateEnforcer::m5g_init_session_credit(SessionMap& session_map,
+		const std::string& imsi, const std::string& session_id,
+	       	const SessionConfig& cfg) {
+   /* creating SessionState object with state CREATING
+    * This calls constructor and allocates memory*/
+   std::cerr << __LINE__ << " " << __FUNCTION__ <<" New SessionState object getting created" 
+	                  << " with IMSI" << imsi <<"\n";
+   auto session_state = std::make_unique<SessionState>
+                        (imsi, session_id, cfg,*rule_store_);
+   MLOG(MINFO) << __LINE__ << " New SessionState object created with IMSI: "
+	        << imsi <<" session context id : " << session_id;
+   handle_session_init_rule_updates(session_map,
+   		                            imsi, *session_state);
+
+   /* Find same UE or imsi already present, if not add
+    * TODO - Need to check if same DNN/APN already exist
+    */
+   auto exist_imsi = session_map.find(imsi);
+   if (exist_imsi == session_map.end()) {
+     // First time a session is created for IMSI in the SessionMap
+     MLOG(MDEBUG) << "First session for IMSI " << imsi
+                  << " with session context ID " << session_id;
+     session_map[imsi] = std::vector<std::unique_ptr<SessionState>>();
+   }
+   else {//TODO IMP to remove the comment, as session_state is not filled properly
+      //session_map[imsi].push_back(std::move(session_state));
+   }
+   /*Check if version missmatch, adapt AMF version as it is new session*/
+   if((session_state->get_current_version()) !=

```

Comment:
If `m5g_init_session_credit` function is solely used for session creation, (when intended state is CREATING), when would we get into the case where the versions differ?

Reply:
Yes, its creating state, but if AMF sends any different version other than 0, SessionD need to adapt. If you think its not required, we can remove this code.

Reply:
Yes I think it would be good to handle the non-CREATING states somewhere else, as the function is called `init_session_credit`. Also, we don't have to use the same naming schemes as the 4G. It might just be cleaner to call the function `initialize_creating_session` or something intuitive like that.

Reply:
Removed and updated.